### PR TITLE
Implement en- and decoding of SYSTEM parameters

### PR DESCRIPTION
**Pull request by andlaus**
If the value of these parameters are not explicitly specified, their value is determined by looking at the current state of the system the program is running on. (If their type is one of the values specified by the ODX standard.)

Andreas Lauser &lt;andreas.lauser@mercedes-benz.com&gt;, on behalf of [MBition GmbH](https://mbition.io/).
[Provider Information](https://github.com/mercedes-benz/foss/blob/master/PROVIDER_INFORMATION.md)

### CODE REVIEW

File: odxtools/compumethods/compucodecompumethod.py [hunk 1]
```diff
@@ -47,17 +47,17 @@ def __post_init__(self) -> None:
...
         return cast(AtomicOdxType, None)
 
     def is_valid_physical_value(self, physical_value: AtomicOdxType) -> bool:
-        odxraise(r"CompuCodeCompuMethod cannot be executed", NotImplementedError)
+        # CompuCodeCompuMethod cannot be executed by odxtools
         return False
 
     def is_valid_internal_value(self, internal_value: AtomicOdxType) -> bool:
-        odxraise(r"CompuCodeCompuMethod cannot be executed", NotImplementedError)
+        # CompuCodeCompuMethod cannot be executed by odxtools
```
Comment (kayoub5) [inline odxtools/compumethods/compucodecompumethod.py:62]: What do the changes in this file have to do with the PR?

File: odxtools/parameterinfo.py [hunk 2]
```diff
@@ -3,15 +3,21 @@
...
 
+from .compumethods.compucodecompumethod import CompuCodeCompuMethod
 from .compumethods.identicalcompumethod import IdenticalCompuMethod
 from .compumethods.limit import IntervalType
 from .compumethods.linearcompumethod import LinearCompuMethod
+from .compumethods.linearsegment import LinearSegment
+from .compumethods.ratfunccompumethod import RatFuncCompuMethod
+from .compumethods.ratfuncsegment import RatFuncSegment
+from .compumethods.scalelinearcompumethod import ScaleLinearCompuMethod
+from .compumethods.scaleratfunccompumethod import ScaleRatFuncCompuMethod
```
Comment (kayoub5) [inline odxtools/parameterinfo.py:14]: Too many changes in this file, that are unrelated to PR topic, please move them to their own PR

File: odxtools/parameters/systemparameter.py [hunk 3]
```diff
@@ -1,19 +1,25 @@
...
+from ..exceptions import odxraise, odxrequire
 from ..odxlink import OdxDocFragment
 from ..odxtypes import ParameterValue
 from ..utils import dataclass_fields_asdict
 from .parameter import ParameterType
 from .parameterwithdop import ParameterWithDOP
 
+PREDEFINED_SYSPARAM_VALUES = [
+    "TIMESTAMP", "SECOND", "MINUTE", "HOUR", "TIMEZONE", "DAY", "WEEK", "MONTH", "YEAR", "CENTURY",
+    "TESTERID", "USERID"
```
Comment (kayoub5) [inline odxtools/parameters/systemparameter.py:20]: Where did this list come from?

File: odxtools/parameters/systemparameter.py [hunk 4]
```diff
@@ -38,18 +44,51 @@ def parameter_type(self) -> ParameterType:
...
+        return True
 
     @override
     def _encode_positioned_into_pdu(self, physical_value: Optional[ParameterValue],
                                     encode_state: EncodeState) -> None:
-        raise NotImplementedError("Encoding a SystemParameter is not implemented yet.")
+        if physical_value is None:
+            # determine the value to be encoded automatically
+            now = datetime.now()
+            if self.sysparam == "TIMESTAMP":
```
Comment (kayoub5) [inline odxtools/parameters/systemparameter.py:63]: Maybe provide users a way to provide a callback for their custom system param, and a way to override default implementation ?

Basically a dict[str, Callback] that users could add to, or change.

File: tests/test_encoding.py [hunk 5]
```diff
@@ -410,6 +414,74 @@ def test_encode_nrc_const(self) -> None:
...
+        )
+        param1 = SystemParameter(
+            oid=None,
+            short_name="year_param",
+            long_name=None,
+            description=None,
+            semantic=None,
+            dop_ref=OdxLinkRef.from_id(dop.odx_id),
+            dop_snref=None,
+            sysparam="YEAR",
```
Comment (kayoub5) [inline tests/test_encoding.py:454]: sysparam or sys_param ?

File: odxtools/compumethods/compucodecompumethod.py (hunk 1 above)
Comment (andlaus) [reply odxtools/compumethods/compucodecompumethod.py:62]: this is in the second commit ("improve `parameter_info()`"). I basically stumbled over this and I thought that the `.is_valid_*_value()` methods should not raise exceptions even if `CompuCode` compu methods cannot be directly supported (as so oftenm I don't have any files which use this attrocity, so I'm pretty much indifferent).

File: odxtools/parameters/systemparameter.py (hunk 3 above)
Comment (andlaus) [reply odxtools/parameters/systemparameter.py:20]: they are listed in table 5 on page 80 of the ASAM spec for ODX 2.2.

File: odxtools/parameters/systemparameter.py (hunk 4 above)
Comment (andlaus) [reply odxtools/parameters/systemparameter.py:63]: if you have a custom sysparam, just specify the value of the parameter explicitly. Adding a mechanism for custom callbacks would add much more complexity for no benefit AFAICS (if the parameter is not predefined, `.is_required` will return true...)

File: tests/test_encoding.py (hunk 5 above)
Comment (andlaus) [reply tests/test_encoding.py:454]: the ODX spec calls the attribute for the `SYSTEM` tag `sysparam`...

File: odxtools/parameterinfo.py (hunk 2 above)
Comment (andlaus) [reply odxtools/parameterinfo.py]: I moved the changes for `free_parameter_info()` into a separate PR: #348

File: odxtools/parameters/systemparameter.py (hunk 3 above)
Comment (kayoub5) [reply odxtools/parameters/systemparameter.py:20]: Maybe add a comment for reference?

**Review by kayoub5 — APPROVED** [review]

File: odxtools/parameters/systemparameter.py (hunk 3 above)
Comment (andlaus) [reply odxtools/parameters/systemparameter.py:24]: fair point. fixed: https://github.com/mercedes-benz/odxtools/pull/347/commits/61c25a8d88c01df37082a9b9f5da3aa3742d8c9f
